### PR TITLE
Remove pkg-config from build process

### LINKED ISSUE #6: Remove pkg_config
The pkg_config dependency is causing problems on Windows and on Mac. On Windows the executable doesn't normally exist at all, and on Mac it's not properly set up for R and thus can't find the right directories.

However, in looking over the code in `build.rs`, I don't see a good reason to use pkg_config at all. It's only use is to find a few directories, and whenever you have a working R install R can do that work for you with the `R.home()` function:

    R.home()
    #> [1] "/Library/Frameworks/R.framework/Resources"
    R.home("lib")
    #> [1] "/Library/Frameworks/R.framework/Resources/lib"
    R.home("include")
    #> [1] "/Library/Frameworks/R.framework/Resources/include"

So my recommendation would be to replace the pkg_config calls with a system call to the R executable, something like the following (but from within Rust, not using the bash command line):

    clauswilke ~> R -s -e 'cat(R.home("include"), "\n")'
    /Library/Frameworks/R.framework/Resources/include 

This should be straightforward with the `std::process` module. If desired, multiple directories can be queried with a single call to R via:

    clauswilke ~> R -s -e 'cat(R.home("include"), R.home("lib"), sep = "\n")'
    /Library/Frameworks/R.framework/Resources/include
    /Library/Frameworks/R.framework/Resources/lib

### PR DESCRIPTION
**Pull request by clauswilke**
This works for me on OS X. I suspect it'll work on Linux and Windows as well, but haven't tested. Would be great if somebody else could.

Closes #6.

### CODE REVIEW

**Comment by CGMossa** [comment]
I've tried it out. Here's what happened:

```
Microsoft Windows [Version 10.0.19041.630]
(c) 2020 Microsoft Corporation. All rights reserved.

C:\Users\tpb398\Documents\GitHub\libR-sys>cargo build
   Compiling winapi v0.3.9
   Compiling memchr v2.3.4
   Compiling cc v1.0.62
   Compiling libc v0.2.80
   Compiling glob v0.3.0
   Compiling proc-macro2 v1.0.24
   Compiling bitflags v1.2.1
   Compiling lazy_static v1.4.0
   Compiling quick-error v1.2.3
   Compiling regex-syntax v0.6.21
   Compiling unicode-width v0.1.8
   Compiling vec_map v0.8.2
   Compiling strsim v0.8.0
   Compiling bindgen v0.53.3
   Compiling peeking_take_while v0.1.2
   Compiling rustc-hash v1.1.0
   Compiling lazycell v1.3.0
   Compiling shlex v0.1.1
   Compiling clang-sys v0.29.3
   Compiling libloading v0.5.2
   Compiling thread_local v1.0.1
   Compiling humantime v1.3.0
   Compiling textwrap v0.11.0
   Compiling log v0.4.11
   Compiling which v3.1.1
   Compiling aho-corasick v0.7.15
   Compiling nom v5.1.2
   Compiling quote v1.0.7
   Compiling regex v1.4.2
   Compiling cexpr v0.4.0
   Compiling atty v0.2.14
   Compiling winapi-util v0.1.5
   Compiling clap v2.33.3
   Compiling termcolor v1.1.0
   Compiling env_logger v0.7.1
   Compiling libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)
error: failed to run custom build command for `libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)`

Caused by:
  process didn't exit successfully: `C:/cargo_target_dir\debug\build\libR-sys-62560aea31121615\build-script-build` (exit code: 1)
  --- stdout
  Problem locating local R instal: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }

C:\Users\tpb398\Documents\GitHub\libR-sys>
```
So I've got an error. 

`R` is not available on my terminal upfront, I then add the path to the 64-bit R via this:
```
SET PATH=%PATH%;C:\R\R-4.0.2\bin\x64
```

A new error occurred:

```
C:\Users\tpb398\Documents\GitHub\libR-sys>cargo build
   Compiling libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)
The following warnings were emitted during compilation:

warning: couldn't execute `llvm-config --prefix` (error: The system cannot find the file specified. (os error 2))
warning: set the LLVM_CONFIG_PATH environment variable to the full path to a valid `llvm-config` executable (including the executable itself)

error: failed to run custom build command for `libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)`

Caused by:
  process didn't exit successfully: `C:/cargo_target_dir\debug\build\libR-sys-62560aea31121615\build-script-build` (exit code: 101)
  --- stdout
  cargo:rustc-env=R_HOME=C:/R/R-4.0.2
  cargo:r_home=C:/R/R-4.0.2
  cargo:rustc-link-search=C:/R/R-4.0.2/lib
  cargo:rustc-link-lib=dylib=R
  cargo:rerun-if-changed=build.rs
  cargo:rerun-if-changed=wrapper.h
  cargo:warning=couldn't execute `llvm-config --prefix` (error: The system cannot find the file specified. (os error 2))  cargo:warning=set the LLVM_CONFIG_PATH environment variable to the full path to a valid `llvm-config` executable (including the executable itself)

  --- stderr
  wrapper.h:9:10: fatal error: 'Rinterface.h' file not found
  wrapper.h:9:10: fatal error: 'Rinterface.h' file not found, err: true
  thread 'main' panicked at 'Unable to generate bindings: ()', build.rs:89:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

So now, I am missing `Rinterface.h`. Which is fine, but I cannot find that myself. 
As I showed before at #6 , it is not there in my default R-installation.

**Comment by clauswilke** [comment]
You're getting the second error because you need to install llvm/libclang and make it available on your path. See here:
https://github.com/rust-lang/rust-bindgen/blob/master/book/src/requirements.md

I had the same error on Mac until I installed llvm.

**Comment by clauswilke** [comment]
This brings up one important point though that'll somehow have to be addressed: Under windows, R packages normally build both 32bit and 64bit executables. This means we'll need two separate versions of `libR-sys`, and the process of querying one executable will not work in this setting. We'd have to know which executable to query for which target.

**Comment by CGMossa** [comment]
I do have llvm installed through scoop
`scoop install llvm`, and it is version 11.0

But I had to set a path to it, i.e. `LIBCLANG_PATH` to `C:\Program Files\LLVM\bin`.

But now even though I did this (and I tried to install LLVM through the installer also), I still get this:
```
   Compiling libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)
error: failed to run custom build command for `libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)`

Caused by:
  process didn't exit successfully: `C:/cargo_target_dir\debug\build\libR-sys-46304186790a078b\build-script-build` (exit code: 101)
  --- stdout
  cargo:rustc-env=R_HOME=C:/R/R-4.0.2
  cargo:r_home=C:/R/R-4.0.2
  cargo:rustc-link-search=C:/R/R-4.0.2/lib
  cargo:rustc-link-lib=dylib=R
  cargo:rerun-if-changed=build.rs
  cargo:rerun-if-changed=wrapper.h

  --- stderr
  wrapper.h:9:10: fatal error: 'Rinterface.h' file not found
  wrapper.h:9:10: fatal error: 'Rinterface.h' file not found, err: true
  thread 'main' panicked at 'Unable to generate bindings: ()', build.rs:89:10
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/cf9cf7c923eb01146971429044f216a3ca905e06\/library\std\src/panicking.rs:495:5
     1: core::panicking::panic_fmt
               at /rustc/cf9cf7c923eb01146971429044f216a3ca905e06\/library\core\src/panicking.rs:92:14
     2: core::option::expect_none_failed
               at /rustc/cf9cf7c923eb01146971429044f216a3ca905e06\/library\core\src/option.rs:1268:5
     3: core::result::Result<T,E>::expect
               at .\C:\Users\tpb398\.rustup\toolchains\nightly-x86_64-pc-windows-gnu\lib/rustlib/src/rust\library\core\src/result.rs:933:23
     4: build_script_build::main
               at .\C:\Users\tpb398\Documents\GitHub\libR-sys/build.rs:86:20
     5: core::ops::function::FnOnce::call_once
               at .\C:\Users\tpb398\.rustup\toolchains\nightly-x86_64-pc-windows-gnu\lib/rustlib/src/rust\library\core\src\ops/function.rs:227:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
But now, it is without the warning about the lack of `llvm-config --prefix`. 

I'd gladly do whatever you guys want me to try on windows, but I'm don't know what to try out myself..

**Comment by clauswilke** [comment]
Apparently `Rinterface.h` does not exist on Windows, and the correct solution is to exclude it from bindgen processing on that platform. See e.g. here:
https://github.com/eddelbuettel/rinside/issues/12#issuecomment-715471988

**Comment by clauswilke** [comment]
From the comments in the source (https://github.com/wch/r-source/blob/dab0c3e8ebcaf375e4a9e882291255f20b19e5a3/src/include/Rinterface.h#L21-L26):

> This header file is to provide hooks for alternative front-ends, e.g. GUIs such as GNOME and Cocoa.  It is only used on Unix-alikes. All entries here should be documented in doc/manual/R-exts.texi. It should not be included by package sources unless they are providing such a front-end.

It probably makes sense to unconditionally exclude it at this time.

**Comment by CGMossa** [comment]
Okay, I've removed it, but now, I'm getting this

```
C:\Users\tpb398\Documents\GitHub\libR-sys(pr/7)> rustup run nightly-gnu cargo build
   Compiling winapi-x86_64-pc-windows-gnu v0.4.0
   Compiling winapi v0.3.9
   Compiling memchr v2.3.4
   Compiling cc v1.0.62
   Compiling version_check v0.9.2
   Compiling glob v0.3.0
   Compiling libc v0.2.80
   Compiling proc-macro2 v1.0.24
   Compiling lazy_static v1.4.0
   Compiling log v0.4.11
   Compiling bitflags v1.2.1
   Compiling regex-syntax v0.6.21
   Compiling cfg-if v0.1.10
   Compiling unicode-xid v0.2.1
   Compiling quick-error v1.2.3
   Compiling unicode-width v0.1.8
   Compiling vec_map v0.8.2
   Compiling bindgen v0.53.3
   Compiling strsim v0.8.0
   Compiling peeking_take_while v0.1.2
   Compiling shlex v0.1.1
   Compiling lazycell v1.3.0
   Compiling rustc-hash v1.1.0
   Compiling nom v5.1.2
   Compiling thread_local v1.0.1
   Compiling clang-sys v0.29.3
   Compiling humantime v1.3.0
   Compiling textwrap v0.11.0
   Compiling libloading v0.5.2
   Compiling aho-corasick v0.7.15
   Compiling which v3.1.1
   Compiling quote v1.0.7
   Compiling regex v1.4.2
   Compiling cexpr v0.4.0
   Compiling atty v0.2.14
   Compiling winapi-util v0.1.5
   Compiling clap v2.33.3
   Compiling termcolor v1.1.0
   Compiling env_logger v0.7.1
   Compiling libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)
error: failed to run custom build command for `libR-sys v0.1.9 (C:\Users\tpb398\Documents\GitHub\libR-sys)`

Caused by:
  process didn't exit successfully: `C:\Users\tpb398\Documents\GitHub\libR-sys\target\debug\build\libR-sys-46304186790a078b\build-script-build` (exit code: 101)
  --- stdout
  cargo:rustc-env=R_HOME=C:/R/R-4.0.2
  cargo:r_home=C:/R/R-4.0.2
  cargo:rustc-link-search=C:/R/R-4.0.2/lib
  cargo:rustc-link-lib=dylib=R
  cargo:rerun-if-changed=build.rs
  cargo:rerun-if-changed=wrapper.h

  --- stderr
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vadefs.h:28:33: error: expected ';' after top level declarator
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:83:1: error: expected identifier
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:83:20: error: expected ';' after top level declarator
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:180:29: error: expected ';' after top level declarator
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:325:51: error: unknown type name 'uintptr_t'
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:326:67: error: unknown type name 'uintptr_t'
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:330:8: error: unknown type name 'uintptr_t'
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:338:1: error: a parameter list without types is only allowed in a function definition
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:12:1: error: expected function body after function declarator
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:287:16: error: unknown type name 'uintptr_t'
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:562:13: error: unknown type name 'size_t'
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:2011:1: error: a parameter list without types is only allowed in a function definition
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_stdio_config.h:12:1: error: expected function body after function declarator
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_stdio_config.h:31:10: error: Unsupported architecture
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:31:14: error: unknown type name 'FILE'
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:49:17: error: unknown type name 'FILE'
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:58:17: error: unknown type name 'FILE'
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:67:17: error: unknown type name 'FILE'
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:79:38: error: unknown type name 'FILE'
  fatal error: too many errors emitted, stopping now [-ferror-limit=]
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vadefs.h:28:33: error: expected ';' after top level declarator, err: true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:83:1: error: expected identifier, err: true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:83:20: error: expected ';' after top level declarator, err: true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:180:29: error: expected ';' after top level declarator, err: true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:325:51: error: unknown type name 'uintptr_t', err: true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:326:67: error: unknown type name 'uintptr_t', err: true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:330:8: error: unknown type name 'uintptr_t', err: true
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include\vcruntime.h:338:1: error: a parameter list without types is only allowed in a function definition, err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:12:1: error: expected function body after function declarator, err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:287:16: error: unknown type name 'uintptr_t', err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:562:13: error: unknown type name 'size_t', err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt.h:2011:1: error: a parameter list without types is only allowed in a function definition, err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_stdio_config.h:12:1: error: expected function body after function declarator, err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_stdio_config.h:31:10: error: Unsupported architecture, err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:31:14: error: unknown type name 'FILE', err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:49:17: error: unknown type name 'FILE', err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:58:17: error: unknown type name 'FILE', err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:67:17: error: unknown type name 'FILE', err: true
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt\corecrt_wstdio.h:79:38: error: unknown type name 'FILE', err: true
  fatal error: too many errors emitted, stopping now [-ferror-limit=], err: true
  thread 'main' panicked at 'Unable to generate bindings: ()', build.rs:89:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
C:\Users\tpb398\Documents\GitHub\libR-sys(pr/7)>
```

I don't understand why it is using windows stuff, as I am running on `nightly-gnu`, which means it should respect that particular toolchain.. I've also ensured to "deactivate" any special settings I might have enabled as part of a `.cargo/config` file.

**Comment by clauswilke** [comment]
I did a quick test. We can't just remove `Rinterface.h`, because it's needed on unix. It's required for this line:
https://github.com/extendr/libR-sys/blob/b9f40d6798ec0e06ca076d7cd4efd5a780964a49/src/lib.rs#L84
and without this line the C stack is too small to call R from Rust.

So it'll probably have to be removed conditionally on Windows.

**Comment by CGMossa** [comment]
Cool.

The current issue I'm experiencing is due to stuff in `Rembedded.h`. 
I thought it might be due to lack of the proper targets, but I have added the right ones like in
the hellorust by jeroen. 

So I'm tinkering still..

**Comment by CGMossa** [comment]
I've gotten a breakthrough: See [issue here if interested](https://github.com/rust-lang/rust-bindgen/issues/1760). `bindgen` defaults to a MSVC toolchain, even though I've configured the main rust-thing to use something else. 

I know get the error that @clauswilke showed above, because that part is not recognised.

Let me tinker a bit more... 

(this is exciting~!)

**Comment by CGMossa** [comment]
I got it to build, but there are many irritating things. 
Assuming that I cannot amend Claus' PR, I'll just list my changes here:
I've added this to  the `wrapper.h`:
```cpp
#include <Rinternals.h>
#ifndef _WIN32
  #define R_INTERFACE_PTRS
  #include <Rinterface.h>
#else
    extern uintptr_t R_CStackLimit; /* C stack limit */
#endif
#include <Rembedded.h>
```
Cut together from `Rinside` and I found out that I need to have `R_CStackLimit` defined to something that is of type usize.

Inside `lib.rs` and around `R_CStackLimit` I've put this:
```rust

            if cfg!(not(target_os = "windows")) {
                R_CStackLimit = usize::max_value();
            }
```

But it is not enough to have a feature flag thing here, one also has to make sure that the variable exist,
so that's why the little macro thing in the wrapper file.

## `build.rs`
Several changes for this one:

I don't know what `R.home("lib")` is supposed to be, as I don't have it on Windows.
I need `R.home("bin")` to get the dynlib R.dll, for instance, and that is located in `R.home("bin")`.

Thus 

```rust
    let rout = Command::new("R")
        .args(&["-s", "-e", r#"cat(R.home(), R.home('include'), R.home('lib'), sep = '\n')"#])
        .args(&[
            "-s",
            "-e",
            r#"cat(R.home(), R.home('include'), R.home('bin'), sep = '\n')"#,
        ])
        .output()?;
```

bindgen defaults to MSVC toolchain, even though we set it differently for the rust library.
Thus the call to bindgen should look like this instead:
```rust
    let bindgen_builder = bindgen_builder.clang_args(&[
        format!("-I{}", &details.include),
        "--target=x86_64-pc-windows-gnu".to_string(),
    ]);
```
Note here that this is for the 64bit thing, for 32 I think it should be 
be `i686-pc-windows-gnu`. 

TODO: I'll try to see if I can inject the chosen one for the current project in this clang-args call.

Note, it is still necessary to add R to one's path to make this work. 

Hooray! Next I'll also see about that `extendR` hello example.

**Comment by clauswilke** [comment]
@CGMossa It's great that you're working through this and finding all these little things that need fixing. Looks like we'll have to do a lot of special casing for Windows.

@andy-thomason Do you have any thoughts on how to best facilitate further development, since it's awkward to have multiple people contribute to one PR? Should we maybe open a branch to which both @CGMossa and I can commit?

**Comment by clauswilke** [comment]
@CGMossa You should be able to extract the correct build target from the Cargo environment variables. See here:
https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts

**Comment by CGMossa** [comment]
Yes I already fixed that. Is that not on my pr?

**Comment by CGMossa** [comment]
It is in line 90 of build.rs. You can just pull from it to yours, and I just kill my pr.

**Comment by clauswilke** [comment]
Yes, you're correct. Sorry, I hadn't noticed you had prepared a PR also.

**Comment by CGMossa** [comment]
You're welcome.
I'm trying to contribute as much as I can. If there is a better workflow to
"submit" commits to PRs, please let me know.

I would just like to know if my changes broke your situation on Mac. Even
though it is already a bit broken due to linker-issues..

By the way, all of this headache is also present in other FFI projects; The
julia bindings are also a pain to setup on Windows for instance..

**Comment by clauswilke** [comment]
In terms of workflow, I can give you access to my forked repo and then you can just push directly into my PR. I just sent you an invite.

**Comment by CGMossa** [comment]
Thanks! I seem to have figured it out. Please report if you still get the "default" result. To test extendr, I had to do the following:

Change the line that "imports" libR-sys in the `extendr-api`'s `Cargo.toml` to look like this:
```toml
libR-sys = {version = "0.1.8", path = "../../libR-sys/"}
```
Because then it looks relative in the path for `libR-sys` package. 

You probably know this, but I'm ensuring that the archives are complete when we are to
write a `troubleshooting.md`, and no-one remembers how we got everything to play nicely.

**Comment by clauswilke** [comment]
Could you add the windows-specific installation instructions (add R to path, install llvm, add to path) to the repo's `README.md` file? I'll check later today whether it still compiles on Mac and will also add the installation instructions.

**Comment by clauswilke** [comment]
I can confirm that the latest patch still compiles and runs correctly on my system.

**Comment by clauswilke** [comment]
I've added general and Mac-specific installation instructions to the README file. @CGMossa Could you fill in the Windows-specific section?

**Comment by CGMossa** [comment]
Yes, I definitely will. Just give me 24 hrs, as it is night time here in
Denmark.

**Comment by CGMossa** [comment]
Added windows instructions.

File: README.md [hunk 1]
```diff
@@ -8,5 +8,92 @@ Low-level R library bindings
...
+2. [libclang](https://clang.llvm.org/docs/Tooling.html): Depending on your operating system, you may need to set the `LIBCLANG_PATH` environment variable or add `llvm-config` to your search path.
+
+### Linux-specific instructions
+
+Set `LIBCLANG_PATH` to the `lib` directory of your `llvm` installation. E.g.,
+`LIBCLANG_PATH=/usr/lib/llvm-3.9/lib`.
+
+### Windows-specific instructions
+
+Ensure that `R_HOME` is an environment variable pointing to the preferred R
```
Comment (clauswilke) [inline README.md:32]: Is the `R_HOME` environment variable actually needed somewhere or can you just set the path directly, without setting `R_HOME`?

**Comment by CGMossa** [comment]
It is not needed anywhere.
At one point, I was going to use it to determine the 32-bit or 64-bit
thing, but it isn't clear how that would work anyways..

Should I remove it? I think it is nicer to know what it is that you're
adding to the path anyways...

**Comment by clauswilke** [comment]
I think it's good practice to keep the installation instructions as simple as possible. So I'd say remove `R_HOME` if it's not needed. You can explain in words what it is you're adding to the path.

File: README.md [hunk 2]
```diff
@@ -8,5 +8,92 @@ Low-level R library bindings
...
+To complete the installation, run `msys2` once, then restart the terminal.
+
+Run `msys2` again, and install Clang and mingw-toolchain via
+
+```bash
+pacman -S --noconfirm mingw-w64-x86_64-clang mingw-w64-x86_64-toolchain
+pacman -S --noconfirm mingw32/mingw-w64-i686-clang mingw-w64-i686-toolchain
+```
+
+Add enivornment variable `LIBCLANG_PATH` with the value pointing to where the
```
Comment (clauswilke) [inline README.md:62]: Note the typo: *environment*

**Comment by clauswilke** [comment]
Oh, one more comment: Your installation instructions will be more accessible if you show the exact lines of code needed to add something to an environment variable, rather than just saying "Add xxx to the environment variable ZZZ."

File: README.md (hunk 2 above)
Comment (CGMossa) [reply README.md]: I've fixed this.

**Comment by CGMossa** [comment]
I've removed `R_HOME` from the instructions.

I don't think it is a good idea to give a command that alters the `PATH`. This is
not straightforward ever, and it is especially error-prone thing on Windows. I've
added a link to some instructions that guide the user, using the Windows UI. That's
the safest way of doing it. Trust me on this one, as just doing this

```
set PATH=%PATH%;"C:/some path/"
```

Is not without its own set of problems. 

Thanks for the feedback

**Comment by clauswilke** [comment]
@andy-thomason Could you review this? Are you willing to merge?

**Comment by andy-thomason** [comment]
All looks pretty good. I hadn't realise how hard you guys were working while I was keeping my
family from killing themselves.

Does this now build on windows?

**Review by andy-thomason — APPROVED** [review]
Give it a whirl. Once we merge, can you test on Windows and I'll then publish the crate.

**Comment by clauswilke** [comment]
I believe @CGMossa had already confirmed this now builds on Windows. But maybe test the version in master one more time to be safe.

**Comment by clauswilke** [comment]
Just to confirm, the version in master builds and tests fine for me on OS X.

**Comment by CGMossa** [comment]
I can also confirm that clean pull from main-branch builds and tests fine for me as well on Windows.

**Comment by CGMossa** [comment]
I've looked at appveyor and windows path variables, and it seems like the solution to add
```
install:
  - cmd: set PATH=C:/R/R-4.0.2/bin/x64;%USERPROFILE%\scoop\apps\msys2\current\mingw64\bin;%USERPROFILE%\scoop\apps\msys2\current\usr\bin;%PATH%
```
See [here](https://help.appveyor.com/discussions/questions/26205-how-do-i-update-the-path-variable-to-the-current-location-of-msbuild15) and [here](https://help.appveyor.com/discussions/problems/14197-adding-linux-and-windows-environment-variable-support) for sources.

For me, the biggest hurdle was being able to determine what "terminal" that should be invoked for this. The Built-in one is `cmd` and this works fine, if everything happens within the session in which this is stet.

**Comment by clauswilke** [comment]
@CGMossa I suggest you create a new issue for setting up an Appveyor build and add your comment regarding windows path variables to that issue. I fear it'll get lost in a comment in a merged PR.

**Pull request merged**